File: backend/internal/models/book.go
package models

import (
	"time"
)

type TabContent struct {
	Video *string `json:"video"`
	Body  string  `json:"body"`
}

// Contains the contents of a book (all book information)
type BookDetails struct {
	ID        string     `json:"id"`
	Title     string     `json:"title"`
	Author    string     `json:"author"`
	Image     *string    `json:"image"` // image link
	Read      TabContent `json:"read"`
	Explore   TabContent `json:"explore"`
	Learn     TabContent `json:"learn"`
	CreatedAt time.Time  `json:"created_at"` // Following ISO 8601
}

// Contains contents of a book without read/explore/learn tabs
type Book struct {
	ID        string    `json:"id"`
	Title     string    `json:"title"`
	Author    string    `json:"author"`
	Image     *string   `json:"image"`      // image link
	Languages []string  `json:"languages"`  // Array of languages (ISO 639-1)
	Favorite  bool      `json:"favorite"`   // if book is favorited
	CreatedAt time.Time `json:"created_at"` // Following ISO 8601
}

type APICreateBook struct {
	Title  string  `json:"title"`
	Author string  `json:"author"`
	Image  *string `json:"image"`
}

type APICreateBookContents struct {
	Language string     `json:"lang"`
	Read     TabContent `json:"read"`
	Explore  TabContent `json:"explore"`
	Learn    TabContent `json:"learn"`
}

type APIUpdateTabContents struct {
	Video *string `json:"video"`
	Body  *string `json:"body"`
}

type APIUpdateBook struct {
	Title  *string `json:"title"`
	Author *string `json:"author"`
	Image  *string `json:"image"`
}

type APIUpdateBookDetails struct {
	Read    APIUpdateTabContents `json:"read"`
	Explore APIUpdateTabContents `json:"explore"`
	Learn   APIUpdateTabContents `json:"learn"`
}

File: web/src/models/Book.ts
import { Language } from './Languages';

export type Book = {
  id: string
  title: string
  author: string
  image?: string
  languages: Language[]
  created_at: string
};

export type BookUpdate = {
  title?: string;
  author?: string;
};

export type BookDetails = {
  id: string
  title: string
  author: string
  image?: string
  read: TabContent
  explore: TabContent
  learn: TabContent
  created_at: string
};

export type TabContent = {
  video?: string;
  body: string;
};

// Nullable TabContent
export type TabContentUpdate = { [K in keyof TabContent]: TabContent[K] | undefined };
